README.md
# koa-required-decorator

一个校验 koa context 特定参数是否存在类方法装饰器 - Check Koa Context parameters exist or not.

## 安装

```bash
npm i koa-required-decorator --save
```

## 使用

跟 `koa-router` 使用的示例，

```javascript
import Required from "koa-required-decorator";

class Controller {
  @Required({
    query: ["id", "name"]
  })
  async getMethod(ctx, _) {
    ctx.body = "GET success";
  }

  @Required({
    "request.fields": ["id", "name"]
  })
  async postMethod(ctx, _) {
    ctx.body = "POST success";
  }
}

const transfer = async (ctx, next) => {
  ctx.request.fields = ctx.query;
  await next();
};

router.get("/demo1", ctrl.getMethod);
router.post("/demo2", transfer, ctrl.postMethod);
```

GET /demo1?id=1

输出结果：

```javascript
{ "errCode": 400000, "errMsg": "parameter \"name\" should not be null." }
```

POST /demo2?name=1

输出结果：

```javascript
{ "errCode": 400000, "errMsg": "parameter \"id\" should not be null." }
```

## API

### @Required(params: IParams)

```
interface IParams {
  [key: string]: string[]
}
```

根据传入的 KEY，查询 `Koa.context` 的特定字段是否含有 VALUE 里列举的属性。<br />
校验对象路径过深，需 `.` 隔开, 如校验 `ctx.request.fields.app.apple`，可以这么配置，

```javascript
@Required({
    'request.fields.app': ['apple']
})
```

### setKoaRequiredOptions(params: IOptions)

```
interface IOptions {
  errMsgFunc?: (errors: string[]) => string;
  responseFunc?: (params: string) => any;
}
```

可以自定义报错文案和返回的数据格式

| 参数         | 说明                 | 类型                                                                                                  | 默认                                                                         |
| ------------ | -------------------- | ----------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------- |
| errMsgFunc   | 定义报错文案         | function(errors: string[]): string<br /> 传参 errors 为被校验为空的字段名集合，出参为定义的文案字符串 | errors => errors.map(k => 'parameter "\${k}" should not be null.').join(' ') |
| responseFunc | 定义报错返回数据格式 | function(params: string): any<br /> 其中 params 为 errMsgFunc 返回的字符串，出参为定义的数据格式      | errMsg => ({errCode: 400000, errMsg: params})                                |

```javascript
import { setKoaRequiredOptions } from "koa-required-decorator";

setKoaRequiredOptions({
  errMsgFunc: arr => arr.map(k => `参数 "${k}" 不能为空.`).join(" "),
  responseFunc: errMsg => {
    return {
      code: 400001,
      msg: errMsg
    };
  }
});
```

GET /demo1

输出结果：

```javascript
{ "code": 400001, "msg": "参数 \"id\" 不能为空. 参数 \"name\" 不能为空." }
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.0.2"></a>
## 0.0.2 (2019-07-24)


__tests__/index.ts
import * as R from 'ramda';
import * as Koa from 'koa';
import * as superagent from 'superagent';
import * as KoaRouter from 'koa-router';
import Required from '../lib';

const app = new Koa();
const router = new KoaRouter();
class Controller {
  @Required({
    query: ['id', 'name']
  })
  async getMethod(ctx, _) {
    ctx.body = 'GET success';
  }

  @Required({
    'request.fields': ['id', 'name']
  })
  async postMethod(ctx, _) {
    ctx.body = 'POST success';
  }

  getReq(url) {
    return new Promise(resolve => {
      superagent.get(`http://localhost:4000${url}`, (_, response) => {
        if (R.has('body')(response) && !R.isNil(response.body) && !R.isEmpty(response.body)) return resolve(response.body);
        resolve(response.text);
      })
    });
  }
  postReq(url) {
    return new Promise(resolve => {
      superagent.post(`http://localhost:4000${url}`, (_, response) => {
        if (R.has('body')(response) && !R.isNil(response.body) && !R.isEmpty(response.body)) return resolve(response.body);
        resolve(response.text);
      })
    });
  }
}
const ctrl = new Controller();
router.get('/demo1', ctrl.getMethod);
router.post('/demo2', async (ctx, next) => {
  ctx.request.fields = ctx.query;
  await next();
}, ctrl.postMethod);

app
  .use(router.routes())
  .use(router.allowedMethods());

let server;;

beforeAll(async done => {
  server = app.listen(4000);
  done();
});

afterAll(async done => {
  await server.close();
  done();
});

describe(`使用默认配置, 校验 query: ['id', 'name']`, () => {
  it('能检测到 GET /demo1, id 和 name 不存在', async () => {
    const data = await ctrl.getReq('/demo1');
    expect(data).toStrictEqual({ "errCode": 400000, "errMsg": "parameter \"id\" should not be null. parameter \"name\" should not be null." });
  });
  it('能检测到 GET /demo1?name=1, id 不存在', async () => {
    const data = await ctrl.getReq('/demo1?name=1');
    expect(data).toStrictEqual({ "errCode": 400000, "errMsg": "parameter \"id\" should not be null." });
  });
  it('能检测到 GET /demo1?id=1, name 不存在', async () => {
    const data = await ctrl.getReq('/demo1?id=1');
    expect(data).toStrictEqual({ "errCode": 400000, "errMsg": "parameter \"name\" should not be null." });
  });
  it('能检测到 GET /demo1?id=1&name=1 无异常', async () => {
    const data = await ctrl.getReq('/demo1?id=1&name=1');
    expect(data).toBe(`GET success`);
  });
});

describe(`使用默认配置, 校验 fields.ctrl.getReq: ['id', 'name']`, () => {
  it('能检测到 POST /demo2, id 和 name 不存在', async () => {
    const data = await ctrl.postReq('/demo2');
    expect(data).toStrictEqual({ "errCode": 400000, "errMsg": "parameter \"id\" should not be null. parameter \"name\" should not be null." });
  });
  it('能检测到 POST /demo2?name=1, id 不存在', async () => {
    const data = await ctrl.postReq('/demo2?name=1');
    expect(data).toStrictEqual({ "errCode": 400000, "errMsg": "parameter \"id\" should not be null." });
  });
  it('能检测到 POST /demo2?id=1, name 不存在', async () => {
    const data = await ctrl.postReq('/demo2?id=1');
    expect(data).toStrictEqual({ "errCode": 400000, "errMsg": "parameter \"name\" should not be null." });
  });
  it('能检测到 POST /demo2?id=1&name=1 无异常', async () => {
    const data = await ctrl.postReq('/demo2?id=1&name=1');
    expect(data).toBe(`POST success`);
  });
});


__tests__/custom.ts
import * as R from 'ramda';
import * as Koa from 'koa';
import * as superagent from 'superagent';
import * as KoaRouter from 'koa-router';
import Required, { setKoaRequiredOptions } from '../lib';

setKoaRequiredOptions({
  errMsgFunc: arr => arr.map(k => `参数 "${k}" 不能为空.`).join(' '),
  responseFunc: errMsg => {
    return ({
      code: 400001,
      msg: errMsg
    })
  }
});

const app = new Koa();
const router = new KoaRouter();
class Controller {
  @Required({
    query: ['id', 'name']
  })
  async getMethod(ctx, _) {
    ctx.body = 'GET success';
  }

  @Required({
    'request.fields': ['id', 'name']
  })
  async postMethod(ctx, _) {
    ctx.body = 'POST success';
  }

  getReq(url) {
    return new Promise(resolve => {
      superagent.get(`http://localhost:4000${url}`, (_, response) => {
        if (R.has('body')(response) && !R.isNil(response.body) && !R.isEmpty(response.body)) return resolve(response.body);
        resolve(response.text);
      })
    });
  }
  postReq(url) {
    return new Promise(resolve => {
      superagent.post(`http://localhost:4000${url}`, (_, response) => {
        if (R.has('body')(response) && !R.isNil(response.body) && !R.isEmpty(response.body)) return resolve(response.body);
        resolve(response.text);
      })
    });
  }
}
const ctrl = new Controller();
router.get('/demo1', ctrl.getMethod);
router.post('/demo2', async (ctx, next) => {
  ctx.request.fields = ctx.query;
  await next();
}, ctrl.postMethod);

app
  .use(router.routes())
  .use(router.allowedMethods());

let server;;

beforeAll(async done => {
  server = app.listen(4000);
  done();
});

afterAll(async done => {
  await server.close();
  done();
});

describe(`使用自定义配置,进行校验`, () => {
  it('能检测到 GET /demo1, id 和 name 不存在', async () => {
    const data = await ctrl.getReq('/demo1');
    expect(data).toStrictEqual({ "code": 400001, "msg": "参数 \"id\" 不能为空. 参数 \"name\" 不能为空." });
  });
  it('能检测到 POST /demo2, id 和 name 不存在', async () => {
    const data = await ctrl.postReq('/demo2');
    expect(data).toStrictEqual({ "code": 400001, "msg": "参数 \"id\" 不能为空. 参数 \"name\" 不能为空." });
  });
});


src/index.ts
import * as R from 'ramda';

interface IOptions {
  errMsgFunc?: (errors: string[]) => string;
  responseFunc?: (params: string) => any;
}

interface IParams {
  [key: string]: string[]
}

const collectionFunc = (errors, obj, params) =>
  R.forEachObjIndexed((value: any, key: string) => {
    R.forEach(R.when((k: string) => R.isNil(R.path(key.split('.').concat(k))(obj)), k => errors.push(k)))(value)
  })(params);

const composeErrMsg = R.compose(
  R.join(' '),
  R.map(k => `parameter "${k}" should not be null.`)
);

let defaultOptions = {
  errMsgFunc: composeErrMsg,
  responseFunc: errMsg => {
    return ({
      errCode: 400000,
      errMsg
    })
  }
}

export const setKoaRequiredOptions = (options: IOptions) => {
  if (R.isNil(options) || R.isEmpty(options) || typeof options !== 'object') return;
  defaultOptions = { ...defaultOptions, ...options };
};

export default (params: IParams) => {
  return (_, __, descriptor) => {
    const fn = descriptor.value;
    descriptor.value = async (ctx, next) => {
      const errors = [];
      collectionFunc(errors, ctx, params);
      if (!R.isEmpty(errors)) {
        ctx.status = 400;
        return ctx.body = defaultOptions.responseFunc(defaultOptions.errMsgFunc(errors));
      }
      await fn(ctx, next);
    }
  }
}

lib/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
var R = require("ramda");
var collectionFunc = function (errors, obj, params) {
    return R.forEachObjIndexed(function (value, key) {
        R.forEach(R.when(function (k) { return R.isNil(R.path(key.split('.').concat(k))(obj)); }, function (k) { return errors.push(k); }))(value);
    })(params);
};
var composeErrMsg = R.compose(R.join(' '), R.map(function (k) { return "parameter \"" + k + "\" should not be null."; }));
var defaultOptions = {
    errMsgFunc: composeErrMsg,
    responseFunc: function (errMsg) {
        return ({
            errCode: 400000,
            errMsg: errMsg
        });
    }
};
exports.setKoaRequiredOptions = function (options) {
    if (R.isNil(options) || R.isEmpty(options) || typeof options !== 'object')
        return;
    defaultOptions = __assign({}, defaultOptions, options);
};
exports.default = (function (params) {
    return function (_, __, descriptor) {
        var fn = descriptor.value;
        descriptor.value = function (ctx, next) { return __awaiter(_this, void 0, void 0, function () {
            var errors;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        errors = [];
                        collectionFunc(errors, ctx, params);
                        if (!R.isEmpty(errors)) {
                            ctx.status = 400;
                            return [2 /*return*/, ctx.body = defaultOptions.responseFunc(defaultOptions.errMsgFunc(errors))];
                        }
                        return [4 /*yield*/, fn(ctx, next)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); };
    };
});
